### PR TITLE
ipset: split into a mutable builder and an immutable set.

### LINKED ISSUE #138: Proposal: make IPSet immutable, add IPSetBuilder
Trying to use IPSet in code, it's annoying that it's mutable. It means I have to either defensively copy, or convert into Ranges and reimplement Contains.

Separately, IPSet.ContainsFunc has a todo about maybe using a more efficient representation for pure matching, once the IPSet is constructed.

Empirically, so far all my uses of IPSet have a "build" phase where I add/remove prefixes, followed by a "use" phase where I want read-only use.

How about making a mutable IPSetBuilder, which produces an immutable IPSet? Then we have the freedom to vary IPSet's representation for efficient read ops; IPSet can behave more like the other types in netaddr (no risk of surprise mutations); and we're more explicit about isolating the more expensive operations into a "build" phase. We can have IPSet.ToBuilder(), for the cases where you have an IPSet you like but just want to tweak a bit.

/cc @bradfitz @josharian @mdlayher thoughts?

### PR DESCRIPTION
**Pull request by danderson**
Fixes #138.

### CODE REVIEW

File: ipset.go [hunk 1]
```diff
@@ -4,167 +4,36 @@
 
 package netaddr
 
-import "sort"
-
-// IPSet represents a set of IP addresses.
+// IPSetBuilder is an IPSet builder.
```
Comment (bradfitz) [inline ipset.go:7]: IPSetBuilder builds an immutable IPSet.

?

Less tautological?

File: ipset.go [hunk 2]
```diff
@@ -268,59 +137,224 @@ func (s *IPSet) Ranges() []IPRange {
...
+}
+
+// Add adds ip to the set s.
+func (s *IPSetBuilder) Add(ip IP) { s.AddRange(IPRange{ip, ip}) }
+
+// AddPrefix adds p's range to s.
+func (s *IPSetBuilder) AddPrefix(p IPPrefix) { s.AddRange(p.Range()) }
 
-	return ret
+// AddRange adds r to s.
```
Comment (bradfitz) [inline ipset.go:156]: If r is not Valid, AddRange does nothing.

?

File: ipset.go [hunk 3]
```diff
@@ -268,59 +137,224 @@ func (s *IPSet) Ranges() []IPRange {
...
+	o.RemoveSet(b)
+	s.removeBuilder(&o)
+}
+
+func discardf(format string, args ...interface{}) {}
+
+// debugf is reassigned by tests.
+var debugf = discardf
+
+func (s *IPSetBuilder) IPSet() *IPSet {
```
Comment (bradfitz) [inline ipset.go:257]: doc?

File: ipset.go [hunk 4]
```diff
@@ -268,59 +137,224 @@ func (s *IPSet) Ranges() []IPRange {
...
+	return true
 }
 
-// Equal reports whether s contains exactly the same IPs as t.
-func (s *IPSet) Equal(t *IPSet) bool {
-	sr := s.Ranges()
-	tr := t.Ranges()
-	if len(sr) != len(tr) {
-		return false
+// Contains returns whether ip is in s.
```
Comment (bradfitz) [inline ipset.go:308]: reports

File: ipset.go [hunk 5]
```diff
@@ -268,59 +137,224 @@ func (s *IPSet) Ranges() []IPRange {
...
+func (s *IPSet) Contains(ip IP) bool {
+	for _, r := range s.rr {
```
Comment (bradfitz) [inline ipset.go:310]: // TODO use a better data structure with link to that github issue?

File: ipset.go [hunk 6]
```diff
@@ -268,59 +137,224 @@ func (s *IPSet) Ranges() []IPRange {
...
+func (s *IPSetBuilder) IPSet() *IPSet {
+	s.normalize()
+	return &IPSet{
+		rr: append([]IPRange{}, s.in...),
+	}
+}
+
+// IPSet represents a set of IP addresses.
+//
+// The zero value is a valid value representing a set of no IPs.
```
Comment (bradfitz) [inline ipset.go:266]: Add that it's safe for concurrent use.

File: ipset.go [hunk 7]
```diff
@@ -268,59 +137,224 @@ func (s *IPSet) Ranges() []IPRange {
...
+type IPSet struct {
+	rr []IPRange // Sorted by IPRange.less
+}
+
+// Clone returns a copy of s that shares no memory with s.
+func (s *IPSet) Clone() *IPSet {
```
Comment (bradfitz) [inline ipset.go:272]: But why if it's immutable?

Oh, mention that it's immutable in the type's godoc too?

File: ipset.go [hunk 8]
```diff
@@ -268,59 +137,224 @@ func (s *IPSet) Ranges() []IPRange {
...
+			return true
+		}
 	}
-	for i := range sr {
-		if sr[i] != tr[i] {
-			return false
+	return false
+}
+
+// ContainsRange returns whether all IPs in o are in s.
```
Comment (bradfitz) [inline ipset.go:318]: "reports whether" here and elsewhere

File: ipset.go [hunk 9]
```diff
@@ -4,167 +4,36 @@
...
-// The Add and Remove methods add or remove IPs to/from the set.
-// Add methods should be called first, as a remove operation does
-// nothing on an empty set. Ranges may be fully, partially, or not
-// overlapping.
-type IPSet struct {
+// The Add and Remove methods add or remove IPs to/from the
+// set. Removals only affect the current membership of the set, so in
+// general Adds should be called first. Input ranges may overlap in
+// any way.
+type IPSetBuilder struct {
```
Comment (bradfitz) [inline ipset.go:15]: regarding your name qualms: if this were called MutableIPSet, that would make it obvious that IPSet weren't mutable, and would make `RemoveFreePrefix` less gross (not being a method on a "builder" type)

File: ipset.go [hunk 10]
```diff
@@ -268,59 +137,224 @@ func (s *IPSet) Ranges() []IPRange {
...
+			debugf("min=%v", min)
 		}
 	}
 
-	// TODO: possibly update s.in and s.out, if #110 supports that.
+	s.in = min
+	s.out = nil
+}
+
+// Add adds ip to the set s.
```
Comment (josharian) [inline ipset.go:150]: s/the set //

File: ipset.go [hunk 11]
```diff
@@ -268,59 +137,224 @@ func (s *IPSet) Ranges() []IPRange {
...
+func (s *IPSetBuilder) Add(ip IP) { s.AddRange(IPRange{ip, ip}) }
+
+// AddPrefix adds p's range to s.
```
Comment (josharian) [inline ipset.go:153]: Or maybe clearer

// AddPrefix adds all IPs in p to s.

File: ipset.go [hunk 12]
```diff
@@ -268,59 +137,224 @@ func (s *IPSet) Ranges() []IPRange {
...
+	}
+	// If there are any removals (s.out), then we need to compact the set
+	// first to get the order right.
+	if len(s.out) > 0 {
+		s.normalize()
+	}
+	s.in = append(s.in, r)
+}
+
+// AddSet adds all ranges in b to s.
```
Comment (josharian) [inline ipset.go:169]: s/all ranges/all IPs/

File: ipset.go [hunk 13]
```diff
@@ -268,59 +137,224 @@ func (s *IPSet) Ranges() []IPRange {
...
+func (s *IPSetBuilder) AddSet(b *IPSet) {
+	for _, r := range b.rr {
+		s.AddRange(r)
+	}
+}
+
+// Remove removes ip from the set s.
```
Comment (josharian) [inline ipset.go:176]: s/the set //

File: ipset.go [hunk 14]
```diff
@@ -268,59 +137,224 @@ func (s *IPSet) Ranges() []IPRange {
...
+func (s *IPSetBuilder) Remove(ip IP) { s.RemoveRange(IPRange{ip, ip}) }
+
+// RemoveFreePrefix removes and returns a Prefix of length bits from the IPSetBuilder.
```
Comment (josharian) [inline ipset.go:179]: See my API comment on the corresponding issue.

File: ipset.go [hunk 15]
```diff
@@ -268,59 +137,224 @@ func (s *IPSet) Ranges() []IPRange {
...
+	if bestFit.IP.IsZero() {
+		return IPPrefix{}, false
+	}
+
+	prefix := IPPrefix{IP: bestFit.IP, Bits: bitLen}
+	s.RemovePrefix(prefix)
+	return prefix, true
+}
+
+// RemovePrefix removes p's range from s.
```
Comment (josharian) [inline ipset.go:208]: // RemovePrefix removes all IPs in p from s.

File: ipset.go [hunk 16]
```diff
@@ -268,59 +137,224 @@ func (s *IPSet) Ranges() []IPRange {
...
+func (s *IPSetBuilder) RemovePrefix(p IPPrefix) { s.RemoveRange(p.Range()) }
+
+// RemoveRange removes r from s.
```
Comment (josharian) [inline ipset.go:211]: // RemoveRange removes all IPs in r from s.

File: ipset.go [hunk 17]
```diff
@@ -268,59 +137,224 @@ func (s *IPSet) Ranges() []IPRange {
...
+func (s *IPSetBuilder) RemoveRange(r IPRange) {
+	if r.Valid() {
+		s.out = append(s.out, r)
+	}
+}
+
+// RemoveSet removes all ranges in o from s.
```
Comment (josharian) [inline ipset.go:218]: // RemoveSet removes all IPs in o from s.

Also, why `o` here and `b` above? Let's stick with `b`, because it's less visually ambiguous.

File: ipset.go [hunk 18]
```diff
@@ -268,59 +137,224 @@ func (s *IPSet) Ranges() []IPRange {
...
+
+// removeBuilder removes all IPs in b from s.
+func (s *IPSetBuilder) removeBuilder(b *IPSetBuilder) {
+	b.normalize()
+	for _, r := range b.in {
+		s.RemoveRange(r)
+	}
+}
+
+// Complement updates s to contain the complement of its current
```
Comment (josharian) [inline ipset.go:233]: I thought we were going to add a universe argument to complement? Or at least document what the universe is.

File: ipset.go (hunk 6 above)
Comment (josharian) [reply ipset.go:266]: Add

// Use IPSetBuilder to construct IPSets.

File: ipset.go [hunk 19]
```diff
@@ -268,59 +137,224 @@ func (s *IPSet) Ranges() []IPRange {
...
+	return &IPSet{
+		rr: append([]IPRange{}, s.in...),
+	}
+}
+
+// IPSet represents a set of IP addresses.
+//
+// The zero value is a valid value representing a set of no IPs.
+type IPSet struct {
+	rr []IPRange // Sorted by IPRange.less
```
Comment (josharian) [inline ipset.go:268]: nit: s/Sorted/sorted/

File: ipset.go (hunk 7 above)
Comment (josharian) [reply ipset.go:272]: I can see wanting an IPSetBuilder `Clone` method.

File: ipset.go (hunk 8 above)
Comment (josharian) [reply ipset.go:318]: Can we use `r` for consistency? We can pick a different local variable name (sr? x?); this one is more important.

File: ipset.go [hunk 20]
```diff
@@ -268,59 +137,224 @@ func (s *IPSet) Ranges() []IPRange {
...
+	for _, r := range s.rr {
+		if o.coveredBy(r) {
+			return true
 		}
 	}
-	return true
+	return false
+}
+
+// ContainsPrefix returns whether all IPs in p are in s.
```
Comment (josharian) [inline ipset.go:328]: s/returns/reports/

File: ipset.go [hunk 21]
```diff
@@ -268,59 +137,224 @@ func (s *IPSet) Ranges() []IPRange {
...
+func (s *IPSet) ContainsPrefix(p IPPrefix) bool {
+	return s.ContainsRange(p.Range())
+}
+
+// Overlaps returns whether any IP in o is also in s.
```
Comment (josharian) [inline ipset.go:333]: s/returns/reports/

Comment (josharian) [reply ipset.go:333]: And s/o/b/ (or t if you don't like b)

File: ipset.go [hunk 22]
```diff
@@ -268,59 +137,224 @@ func (s *IPSet) Ranges() []IPRange {
...
+		for _, or := range o.rr {
+			if r.Overlaps(or) {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// OverlapsRange returns whether any IP in o is also in s.
```
Comment (josharian) [inline ipset.go:346]: reports
t or b

File: ipset.go [hunk 23]
```diff
@@ -268,59 +137,224 @@ func (s *IPSet) Ranges() []IPRange {
...
+	for _, r := range s.rr {
+		if r.Overlaps(o) {
+			return true
+		}
+	}
+	return false
+}
+
+// OverlapsPrefix returns whether any IP in p is also in s.
+func (s *IPSet) OverlapsPrefix(p IPPrefix) bool {
```
Comment (josharian) [inline ipset.go:358]: Is it worth having all all these API variants, when it's so easy to get a range out of a prefix? (Maybe it is, but I want to make sure we consider the question.)

**Review by josharian — COMMENTED** [review]
(My GitHub notifications for netaddr go to my Tailscale email address, so expect slow turnarounds from me on any netaddr discussions...unless explicitly mentioned, in which case my phone dings. 🤷🏽 )

File: ipset.go (hunk 2 above)
Comment (danderson) [reply ipset.go:156]: Reminds me I want to maybe change that, the whole "mysteriously do nothing on invalid input" makes me a bit nervous. Filing bug about that.

File: ipset.go (hunk 7 above)
Comment (danderson) [reply ipset.go:272]: Good point, no reason to have Clone here. Moved to IPSetBuilder.

File: ipset.go (hunk 9 above)
Comment (danderson) [reply ipset.go:15]: MutableIPSet suggests that it should have all the various accessors that IPSet has, in my mind. I'd rather avoid that and keep it as a pure builder type.

I'm leaning a little more to Josh's suggestion of moving RemoveFreePrefix over to IPSet, and just make it produce a new IPSet on use. Makes it a little more expensive to use, but it's not a hot path.

File: ipset.go (hunk 18 above)
Comment (danderson) [reply ipset.go:233]: Separate issue. Going to leave this as-is for now, as it's as broken as the previous code.

File: ipset.go (hunk 22 above)
Comment (danderson) [reply ipset.go:346]: Used r, for consistency with other IPRange methods.

File: ipset.go (hunk 23 above)
Comment (danderson) [reply ipset.go:358]: I didn't consider it too hard, honestly. I figured it's cheap for us to provide accessors for all the top-level types in netaddr, rather than require callers to add visual noise.

File: ipset.go (hunk 14 above)
Comment (danderson) [reply ipset.go]: Moved RemoveFreePrefix onto IPSet.

File: ipset.go [hunk 24]
```diff
@@ -268,59 +137,233 @@ func (s *IPSet) Ranges() []IPRange {
...
-	return true
+	return false
+}
+
+// ContainsPrefix reports whether all IPs in p are in s.
+func (s *IPSet) ContainsPrefix(p IPPrefix) bool {
+	return s.ContainsRange(p.Range())
+}
+
+// Overlaps reports whether any IP in o is also in s.
```
Comment (bradfitz) [inline ipset.go:311]: `"in o"` but the arguments are named `s` and `b.

File: ipset.go [hunk 25]
```diff
@@ -268,59 +137,233 @@ func (s *IPSet) Ranges() []IPRange {
...
+		for _, or := range b.rr {
+			if r.Overlaps(or) {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// OverlapsRange reports whether any IP in o is also in s.
```
Comment (bradfitz) [inline ipset.go:324]: same

File: ipset.go [hunk 26]
```diff
@@ -268,59 +137,233 @@ func (s *IPSet) Ranges() []IPRange {
...
+	return false
+}
+
+// OverlapsPrefix reports whether any IP in p is also in s.
+func (s *IPSet) OverlapsPrefix(p IPPrefix) bool {
+	return s.OverlapsRange(p.Range())
+}
+
+// RemoveFreePrefix splits s into a Prefix of length bitLen and a new
+// IPSet with that prefix removed.
```
Comment (bradfitz) [inline ipset.go:341]: worth mentioning what ok==false means? That no such space exists.

File: ipset.go [hunk 27]
```diff
@@ -268,59 +137,233 @@ func (s *IPSet) Ranges() []IPRange {
...
+	}
+}
+
+// Add adds ip to s.
+func (s *IPSetBuilder) Add(ip IP) { s.AddRange(IPRange{ip, ip}) }
+
+// AddPrefix adds all IPs in p to s.
+func (s *IPSetBuilder) AddPrefix(p IPPrefix) { s.AddRange(p.Range()) }
+
+// AddRange adds r to s. If r is not Valid, AddRange does nothing.
```
Comment (bradfitz) [inline ipset.go:164]: If on a new line

Also should the other two Add methods above also say the equivalent thing?

Comment (danderson) [reply ipset.go:164]: The other Add methods take types which cannot be invalid, so no such problem applies.

File: ipset.go [hunk 28]
```diff
@@ -268,59 +137,237 @@ func (s *IPSet) Ranges() []IPRange {
...
-	// with s.Prefixes()?
-	return func(ip IP) bool {
-		i := sort.Search(len(rv), func(i int) bool {
-			return ip.Less(rv[i].From)
-		})
-		if i == 0 {
+// Equal reports whether s and o represent the same set of IP
+// addresses.
+func (s *IPSet) Equal(o *IPSet) bool {
+	if len(s.rr) != len(o.rr) {
```
Comment (bradfitz) [inline ipset.go:274]: this implementation assumes that s.rr is also normalized; add a comment to line 252 where it says "sorted by IPRange.less" to say a bit more? (might need to promote it to a not-same-line comment)

File: ipset.go [hunk 29]
```diff
@@ -268,59 +137,237 @@ func (s *IPSet) Ranges() []IPRange {
...
 
-// ContainsFunc returns a func that reports whether an IP is in s.
-// The returned func operates on a copy of s, so s may be mutated
-// later.
-func (s *IPSet) ContainsFunc() (contains func(IP) bool) {
-	rv := s.Ranges()
-	// TODO(bradfitz): build a faster data structure with
-	// with s.Prefixes()?
-	return func(ip IP) bool {
-		i := sort.Search(len(rv), func(i int) bool {
```
Comment (bradfitz) [inline ipset.go:302]: why did you delete this? This was a better implementation of Contains, no?

**Review by bradfitz — APPROVED** [review]

File: ipset.go (hunk 28 above)
Comment (danderson) [reply ipset.go:274]: Documented.

File: ipset.go (hunk 29 above)
Comment (danderson) [reply ipset.go:302]: I deleted ContainsFunc because we now have a plain Contains. Not porting the implementation over was an oversight. Fixed.
